packages/babel-plugin-minify-simplify/src/index.js

```diff
@@ -124,6 +124,54 @@ module.exports = ({ types: t }) => {
         ],
       },
 
+      LogicalExpression(path) {
+        const left = path.get("left");
+        const right = path.get("right");
+        const operator = path.node.operator;
+
+        const AND = (input) => input === "&&";
+        const OR = (input) => input === "||";
+
+        const TRUTHY = (input) => {
+          const evalResult = input.evaluate();
+          return evalResult.confident && input.isPure() && evalResult.value;
+        };
+        const FALSY = (input) => {
+          // NaN and undefined are falsy
+          if (input.isIdentifier()) {
+            if (input.node.name === "NaN" || input.node.name === "undefined") {
+              return true;
+            }
+          }
+          const evalResult = input.evaluate();
+          return evalResult.confident && input.isPure() && !evalResult.value;
+        };
+
+        const {
+          Expression: EX
+        } = types;
+
+        // Convention:
+        // [left, operator, right, handler(leftNode, rightNode)]
+        const matcher = new PatternMatch([
+          [TRUTHY, AND, EX, (l, r) => r],
+          [FALSY, AND, EX, (l) => l],
+          [TRUTHY, OR, EX, (l) => l],
+          [FALSY, OR, EX, (l, r) => r]
+        ]);
+
+        const result = matcher.match(
+          [left, operator, right],
+          isPatternMatchesPath
+        );
+
+        if (result.match) {
+          // here we are sure that left.evaluate is always confident becuase
+          // it satisfied one of TRUTHY/FALSY paths
+          path.replaceWith(result.value(t.valueToNode(left.evaluate().value), right.node));
```

Comment:
Would it be worth caching/re-using the result of `left.evaluate()`? I don't know how expensive these evaluations generally are.